src/lib.rs

```diff
@@ -91,3 +92,95 @@ pub fn format_size(size: u64, binary: bool) -> String {
         NumberPrefix::Prefixed(prefix, number) => format!("{:.2} {}B", number, prefix),
     }
 }
+
+/// Calculate the sum of sizes of all entries
+///
+/// Does not account twice the size of entries that are within other entries.
+/// Implementation uses BTreeMap to build a path Trie.
+pub fn calculate_unique_total_size(entries: &[Entry]) -> u64 {
+    let mut filtered_entries = Vec::<&Entry>::new();
+    let mut canonicalized_paths = Vec::<PathBuf>::new();
+
+    // Canonicalize each path, silently ignoring failures.
+    // TODO: Review if we should ignore failures.
+    for entry in entries {
+        if let Ok(path) = entry.path.canonicalize() {
+            filtered_entries.push(entry);
+            canonicalized_paths.push(path);
+        }
+    }
+
+    #[derive(PartialEq, Eq, PartialOrd, Ord)]
+    struct TriePathNode {
+        // Children nodes of this current path, accessed by path.
+        children: BTreeMap<PathBuf, TriePathNode>,
+        // Size of the file that ends at this node.
+        node_size: u64,
+    }
+
+    let mut trie_root = TriePathNode {
+        children: BTreeMap::new(),
+        node_size: 0,
+    };
+
+    // For each entry/path, add it to the Trie if it wasn't already inserted.
+    //
+    // If the Trie receives a folder that is parent of a previously added file, then just consider
+    // the parent folder, removing the childs, this way, we do not count them twice towards the
+    // final total.
+    for (i, entry) in filtered_entries.iter().enumerate() {
+        let path = &canonicalized_paths[i];
```

Comment:
Consider zipping those two sequences instead of enumerating and indexing.

Reply:
Done!